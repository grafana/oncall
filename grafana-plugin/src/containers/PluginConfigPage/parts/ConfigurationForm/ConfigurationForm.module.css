.info-block {
  margin-bottom: 24px;
  margin-top: 24px;
}
