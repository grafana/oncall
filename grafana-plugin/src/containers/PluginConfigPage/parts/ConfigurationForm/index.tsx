import React, { FC, useCallback, useState } from 'react';

import { Button, Field, Form, Input } from '@grafana/ui';
import cn from 'classnames/bind';
import { isEmpty } from 'lodash-es';
import { SubmitHandler } from 'react-hook-form';

import Block from 'components/GBlock/Block';
import Text from 'components/Text/Text';
import PluginState from 'state/plugin';
import { getItem, ONCALL_API_URL_LOCAL_STORAGE_KEY, setItem } from 'utils/localStorage';

import styles from './ConfigurationForm.module.css';

const cx = cn.bind(styles);

type Props = {
  onSuccessfulSetup: () => void;
};

type FormProps = {
  onCallApiUrl: string;
};

/**
 * https://stackoverflow.com/a/43467144
 */
const isValidUrl = (url: string): boolean => {
  try {
    new URL(url);
    return true;
  } catch (_) {
    return false;
  }
};

const FormErrorMessage: FC<{ errorMsg: string }> = ({ errorMsg }) => (
  <>
    <pre>
      <Text type="link">{errorMsg}</Text>
    </pre>
    <Block withBackground className={cx('info-block')}>
      <Text type="secondary">
        Need help?
        <br />- Reach out to the OnCall team in the{' '}
        <a href="https://grafana.slack.com/archives/C02LSUUSE2G" target="_blank" rel="noreferrer">
          <Text type="link">#grafana-oncall</Text>
        </a>{' '}
        community Slack channel
        <br />- Ask questions on our GitHub Discussions page{' '}
        <a href="https://github.com/grafana/oncall/discussions/categories/q-a" target="_blank" rel="noreferrer">
          <Text type="link">here</Text>
        </a>{' '}
        <br />- Or file bugs on our GitHub Issues page{' '}
        <a href="https://github.com/grafana/oncall/issues" target="_blank" rel="noreferrer">
          <Text type="link">here</Text>
        </a>
      </Text>
    </Block>
  </>
);

const ConfigurationForm: FC<Props> = ({ onSuccessfulSetup }) => {
  const [setupErrorMsg, setSetupErrorMsg] = useState<string>(null);
  const [formLoading, setFormLoading] = useState<boolean>(false);

  const setupPlugin: SubmitHandler<FormProps> = useCallback(async ({ onCallApiUrl }) => {
    setFormLoading(true);

    const errorMsg = await PluginState.selfHostedInstallPlugin(onCallApiUrl);

    if (!errorMsg) {
      setItem(ONCALL_API_URL_LOCAL_STORAGE_KEY, onCallApiUrl);
      onSuccessfulSetup();
    } else {
      setSetupErrorMsg(errorMsg);
      setFormLoading(false);
    }
  }, []);

  return (
    <Form<FormProps>
      defaultValues={{ onCallApiUrl: getItem(ONCALL_API_URL_LOCAL_STORAGE_KEY) }}
      onSubmit={setupPlugin}
      data-testid="plugin-configuration-form"
    >
      {({ register, errors }) => (
        <>
          <div className={cx('info-block')}>
            <p>1. Launch the OnCall backend</p>
            <Text type="secondary">
              Run hobby, dev or production backend. See{' '}
              <a href="https://github.com/grafana/oncall#getting-started" target="_blank" rel="noreferrer">
                <Text type="link">here</Text>
              </a>{' '}
              on how to get started.
            </Text>
          </div>

          <div className={cx('info-block')}>
            <p>2. Let us know the base URL of your OnCall API</p>
            <Text type="secondary">
              The OnCall backend must be reachable from your Grafana installation. Some examples are:
              <br />
              - http://host.docker.internal:8080
              <br />- http://localhost:8080
            </Text>
          </div>

          <Field label="OnCall backend URL" invalid={!!errors.onCallApiUrl} error="Must be a valid URL">
            <Input
              data-testid="onCallApiUrl"
              {...register('onCallApiUrl', {
                required: true,
                validate: isValidUrl,
              })}
            />
          </Field>

          {setupErrorMsg && <FormErrorMessage errorMsg={setupErrorMsg} />}

          <Button type="submit" size="md" disabled={formLoading || !isEmpty(errors)}>
            Connect
          </Button>
        </>
      )}
    </Form>
  );
};

export default ConfigurationForm;
