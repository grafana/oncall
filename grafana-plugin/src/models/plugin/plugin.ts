import { isEqual } from 'lodash-es';
import { makeAutoObservable, runInAction } from 'mobx';
import { OnCallPluginMetaJSONData } from 'types';

import { ActionKey } from 'models/loader/action-keys';
import { GrafanaApiClient } from 'network/grafana-api/http-client';
import { makeRequest } from 'network/network';
import { PluginConnection, StatusResponse } from 'network/oncall-api/api.types';
import { RootBaseStore } from 'state/rootBaseStore/RootBaseStore';
import { waitInMs } from 'utils/async';
import { AutoLoadingState } from 'utils/decorators';

import { PluginHelper } from './plugin.helper';

/* 
High-level OnCall initialization process:
On OSS:
  - On OnCall page / OnCall extension mount POST /status is called and it has pluginConfiguration object with different flags. 
    If all of them have `ok: true` , we consider plugin to be successfully configured and application loading is being continued. 
    Otherwise, we show error page with the option to go to plugin config (for Admin user) or to contact administrator (for nonAdmin user)
  - On plugin config page frontend sends another POST /status. If every flag has `ok: true`, it shows that plugin is connected. 
    Otherwise, it shows more detailed information of what is misconfigured / missing. User can update onCallApiUrl and try to reconnect plugin.
      - If Grafana version >= 10.3 AND externalServiceAccount feature flag is `true`, then grafana token is autoprovisioned and there is no need to create it
      - Otherwise, user is given the option to manually create service account as Admin and then reconnect the plugin
On Cloud:
  - On OnCall page / OnCall extension mount POST /status is called. If plugin is configured correctly, application loads as usual.
    If it's not, we show error page with the button to contact support
  - On plugin config page we show info if plugin is connected. If it's not we show detailed information of the errors and the button to contact support
*/

export class PluginStore {
  rootStore: RootBaseStore;
  connectionStatus?: PluginConnection;
  apiUrlFromStatus?: string;
  isPluginConnected = false;
  appliedOnCallApiUrl = '';

  constructor(rootStore: RootBaseStore) {
    makeAutoObservable(this, undefined, { autoBind: true });
    this.rootStore = rootStore;
  }

  private resetConnectionStatus() {
    this.connectionStatus = undefined;
    this.isPluginConnected = false;
  }

  async refreshAppliedOnCallApiUrl() {
    const { jsonData } = await GrafanaApiClient.getGrafanaPluginSettings();
    runInAction(() => {
      this.appliedOnCallApiUrl = jsonData.onCallApiUrl;
    });
  }

  @AutoLoadingState(ActionKey.PLUGIN_VERIFY_CONNECTION)
  async verifyPluginConnection() {
    const { pluginConnection, api_url } = await makeRequest<StatusResponse>(`/plugin/status`, {});
    runInAction(() => {
      this.connectionStatus = pluginConnection;
      this.apiUrlFromStatus = api_url;
      this.isPluginConnected = Object.keys(pluginConnection).every(
        (key) => pluginConnection[key as keyof PluginConnection]?.ok
      );
    });
  }

  @AutoLoadingState(ActionKey.PLUGIN_UPDATE_SETTINGS_AND_REINITIALIZE)
  async updatePluginSettingsAndReinitializePlugin({
    currentJsonData,
    newJsonData,
  }: {
    currentJsonData: OnCallPluginMetaJSONData;
    newJsonData: Partial<OnCallPluginMetaJSONData>;
  }) {
    this.resetConnectionStatus();
    const saveJsonDataCandidate = { ...currentJsonData, ...newJsonData };
    if (!isEqual(currentJsonData, saveJsonDataCandidate) || !this.connectionStatus?.oncall_api_url?.ok) {
      await GrafanaApiClient.updateGrafanaPluginSettings({ jsonData: saveJsonDataCandidate });
      await waitInMs(1000); // It's required for backend proxy to pick up new settings
    }
    try {
      await PluginHelper.install();
    } finally {
      await this.verifyPluginConnection();
    }
  }

  @AutoLoadingState(ActionKey.PLUGIN_RECREATE_SERVICE_ACCOUNT)
  async recreateServiceAccountAndRecheckPluginStatus() {
    await GrafanaApiClient.recreateGrafanaTokenAndSaveInPluginSettings();
    await this.verifyPluginConnection();
  }

  async enablePlugin() {
    await GrafanaApiClient.updateGrafanaPluginSettings({}, true);
    location.reload();
  }
}
