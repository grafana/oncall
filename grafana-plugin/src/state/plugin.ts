import { getBackendSrv } from '@grafana/runtime';

import { makeRequest } from 'network';

export async function createGrafanaToken() {
  const keys = await getBackendSrv().get('/api/auth/keys');
  const existingKey = keys.find((key: { id: number; name: string; role: string }) => key.name === 'OnCall');

  if (existingKey) {
    await getBackendSrv().delete(`/api/auth/keys/${existingKey.id}`);
  }

  return await getBackendSrv().post('/api/auth/keys', {
    name: 'OnCall',
    role: 'Admin',
    secondsToLive: null,
  });
}

export async function updateGrafanaToken(key: string) {
  await getBackendSrv().post(`/api/plugins/grafana-oncall-app/settings`, {
    enabled: true,
    pinned: true,
    secureJsonData: {
      grafanaToken: key,
    },
  });
}

export async function startPluginSync() {
  return await makeRequest('/plugin/sync', { method: 'POST' });
}

export const SYNC_STATUS_RETRY_LIMIT = 10;

export const syncStatusDelay = (retryCount) => new Promise((resolve) => setTimeout(resolve, 10 * 2 ** retryCount));

export async function getPluginSyncStatus() {
  return await makeRequest(`/plugin/sync`, { method: 'GET' });
}

export async function installPlugin() {
  const grafanaToken = await createGrafanaToken();
  await updateGrafanaToken(grafanaToken.key);
  return await makeRequest('/plugin/install', { method: 'POST' });
}
