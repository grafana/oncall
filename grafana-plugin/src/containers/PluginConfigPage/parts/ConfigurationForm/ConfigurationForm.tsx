import React, { FC, useCallback, useState } from 'react';

import { Button, Field, Form, Input } from '@grafana/ui';
import cn from 'classnames/bind';
import { isEmpty } from 'lodash-es';
import { SubmitHandler } from 'react-hook-form';

import { Block } from 'components/GBlock/Block';
import { Text } from 'components/Text/Text';
import { PluginState } from 'state/plugin/plugin';

import styles from './ConfigurationForm.module.css';

const cx = cn.bind(styles);

type Props = {
  onSuccessfulSetup: () => void;
  defaultOnCallApiUrl: string;
};

type FormProps = {
  onCallApiUrl: string;
};

/**
 * https://stackoverflow.com/a/43467144
 */
const isValidUrl = (url: string): boolean => {
  try {
    new URL(url);
    return true;
  } catch (_) {
    return false;
  }
};

const FormErrorMessage: FC<{ errorMsg: string }> = ({ errorMsg }) => (
  <>
    <pre>
      <Text type="link">{errorMsg}</Text>
    </pre>
    <Block withBackground className={cx('info-block')}>
      <Text type="secondary">
        Need help?
        <br />- Reach out to the OnCall team in the{' '}
        <a href="https://grafana.slack.com/archives/C02LSUUSE2G" target="_blank" rel="noreferrer">
          <Text type="link">#grafana-oncall</Text>
        </a>{' '}
        community Slack channel
        <br />- Ask questions on our GitHub Discussions page{' '}
        <a href="https://github.com/grafana/oncall/discussions/categories/q-a" target="_blank" rel="noreferrer">
          <Text type="link">here</Text>
        </a>{' '}
        <br />- Or file bugs on our GitHub Issues page{' '}
        <a href="https://github.com/grafana/oncall/issues" target="_blank" rel="noreferrer">
          <Text type="link">here</Text>
        </a>
      </Text>
    </Block>
  </>
);

export const ConfigurationForm: FC<Props> = ({ onSuccessfulSetup, defaultOnCallApiUrl }) => {
  const [setupErrorMsg, setSetupErrorMsg] = useState<string>(null);
  const [formLoading, setFormLoading] = useState<boolean>(false);

  const setupPlugin: SubmitHandler<FormProps> = useCallback(async ({ onCallApiUrl }) => {
    setFormLoading(true);

    const errorMsg = await PluginState.selfHostedInstallPlugin(onCallApiUrl, false);

    if (!errorMsg) {
      onSuccessfulSetup();
    } else {
      setSetupErrorMsg(errorMsg);
      setFormLoading(false);
    }
  }, []);

  return (
    <Form<FormProps>
      defaultValues={{ onCallApiUrl: defaultOnCallApiUrl }}
      onSubmit={setupPlugin}
      data-testid="plugin-configuration-form"
    >
      {({ register, errors }) => (
        <>
          <div className={cx('info-block')}>
            <p>1. Launch the OnCall backend</p>
            <Text type="secondary">
              Run hobby, dev or production backend. See{' '}
              <a href="https://github.com/grafana/oncall#getting-started" target="_blank" rel="noreferrer">
                <Text type="link">here</Text>
              </a>{' '}
              on how to get started.
            </Text>
          </div>

          <div className={cx('info-block')}>
            <p>2. Let us know the base URL of your OnCall API</p>
            <Text type="secondary">
              The OnCall backend must be reachable from your Grafana installation. Some examples are:
              <br />
              - http://host.docker.internal:8080
              <br />- http://localhost:8080
            </Text>
          </div>

          <Field label="OnCall backend URL" invalid={!!errors.onCallApiUrl} error="Must be a valid URL">
            <Input
              data-testid="onCallApiUrl"
              {...register('onCallApiUrl', {
                required: true,
                validate: isValidUrl,
              })}
            />
          </Field>

          {setupErrorMsg && <FormErrorMessage errorMsg={setupErrorMsg} />}

          <Button type="submit" size="md" disabled={formLoading || !isEmpty(errors)}>
            Connect
          </Button>
        </>
      )}
    </Form>
  );
};
