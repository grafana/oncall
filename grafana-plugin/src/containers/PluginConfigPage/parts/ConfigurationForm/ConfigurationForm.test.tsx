import React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { PluginState } from 'state/plugin/plugin';

import { ConfigurationForm } from './ConfigurationForm';

jest.mock('state/plugin/plugin');

const VALID_ONCALL_API_URL = 'http://host.docker.internal:8080';
const SELF_HOSTED_PLUGIN_API_ERROR_MSG = 'ohhh nooo there was an error from the OnCall API';

const fillOutFormAndTryToSubmit = async (onCallApiUrl: string, selfHostedInstallPluginSuccess = true) => {
  // mocks
  const mockOnSuccessfulSetup = jest.fn();
  PluginState.selfHostedInstallPlugin = jest
    .fn()
    .mockResolvedValueOnce(selfHostedInstallPluginSuccess ? null : SELF_HOSTED_PLUGIN_API_ERROR_MSG);

  // setup
  const component = render(
    <ConfigurationForm onSuccessfulSetup={mockOnSuccessfulSetup} defaultOnCallApiUrl="http://potato.com" />
  );

  // fill out onCallApiUrl input
  const input = screen.getByTestId('onCallApiUrl');

  await userEvent.click(input);
  await userEvent.clear(input); // clear the input first before typing to wipe out the placeholder text
  await userEvent.keyboard(onCallApiUrl);

  // submit form
  await userEvent.click(screen.getByRole('button'));

  return { dom: component.baseElement, mockOnSuccessfulSetup };
};

describe('ConfigurationForm', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  test('it sets the default input value of onCallApiUrl to the passed in prop value of defaultOnCallApiUrl', () => {
    const processEnvOnCallApiUrl = 'http://hello.com';
    render(<ConfigurationForm onSuccessfulSetup={jest.fn()} defaultOnCallApiUrl={processEnvOnCallApiUrl} />);
    expect(screen.getByDisplayValue(processEnvOnCallApiUrl)).toBeInTheDocument();
  });

  test('It calls the onSuccessfulSetup callback on successful form submission', async () => {
    const { mockOnSuccessfulSetup } = await fillOutFormAndTryToSubmit(VALID_ONCALL_API_URL);

    expect(PluginState.selfHostedInstallPlugin).toHaveBeenCalledTimes(1);
    expect(PluginState.selfHostedInstallPlugin).toHaveBeenCalledWith(VALID_ONCALL_API_URL, false);
    expect(mockOnSuccessfulSetup).toHaveBeenCalledTimes(1);
  });

  test("It doesn't allow the user to submit if the URL is invalid", async () => {
    const { dom, mockOnSuccessfulSetup } = await fillOutFormAndTryToSubmit('potato');

    expect(PluginState.selfHostedInstallPlugin).toHaveBeenCalledTimes(0);
    expect(mockOnSuccessfulSetup).toHaveBeenCalledTimes(0);
    expect(screen.getByRole('button')).toBeDisabled();
    expect(dom).toMatchSnapshot();
  });

  test('It shows an error message if the self hosted plugin API call fails', async () => {
    const { dom, mockOnSuccessfulSetup } = await fillOutFormAndTryToSubmit(VALID_ONCALL_API_URL, false);

    expect(PluginState.selfHostedInstallPlugin).toHaveBeenCalledWith(VALID_ONCALL_API_URL, false);
    expect(mockOnSuccessfulSetup).toHaveBeenCalledTimes(0);
    expect(dom).toMatchSnapshot();
  });
});
