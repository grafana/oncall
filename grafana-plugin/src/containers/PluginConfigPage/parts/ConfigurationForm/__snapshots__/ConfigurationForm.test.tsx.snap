// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigurationForm It doesn't allow the user to submit if the URL is invalid 1`] = `
<body>
  <div>
    <form
      class="css-1srg48i"
      data-testid="plugin-configuration-form"
    >
      <div
        class="info-block"
      >
        <p>
          1. Launch the OnCall backend
        </p>
        <span
          class="root text text--secondary text--medium"
        >
          Run hobby, dev or production backend. See
           
          <a
            href="https://github.com/grafana/oncall#getting-started"
            rel="noreferrer"
            target="_blank"
          >
            <span
              class="root text text--link text--medium"
            >
              here
            </span>
          </a>
           
          on how to get started.
        </span>
      </div>
      <div
        class="info-block"
      >
        <p>
          2. Let us know the base URL of your OnCall API
        </p>
        <span
          class="root text text--secondary text--medium"
        >
          The OnCall backend must be reachable from your Grafana installation. Some examples are:
          <br />
          - http://host.docker.internal:8080
          <br />
          - http://localhost:8080
        </span>
      </div>
      <div
        class="css-1rplq84"
      >
        <div
          class="css-l4ykjo-Label"
        >
          <label>
            <div
              class="css-70qvj9"
            >
              OnCall backend URL
            </div>
          </label>
        </div>
        <div>
          <div>
            <div
              class="css-1skhtb9-input-wrapper"
              data-testid="input-wrapper"
            >
              <div
                class="css-10lnb82-input-inputWrapper"
              >
                <input
                  class="css-11mwy6h-input-input"
                  data-testid="onCallApiUrl"
                  name="onCallApiUrl"
                />
              </div>
            </div>
          </div>
          <div
            class="css-1fhgjcy"
          >
            <div
              class="css-9z7wq3"
              role="alert"
            >
              <div
                class="css-1j2891d-Icon"
              />
              Must be a valid URL
            </div>
          </div>
        </div>
      </div>
      <button
        class="css-td06pi-button"
        disabled=""
        type="submit"
      >
        <span
          class="css-1riaxdn"
        >
          Connect
        </span>
      </button>
    </form>
  </div>
</body>
`;

exports[`ConfigurationForm It shows an error message if the self hosted plugin API call fails 1`] = `
<body>
  <div>
    <form
      class="css-1srg48i"
      data-testid="plugin-configuration-form"
    >
      <div
        class="info-block"
      >
        <p>
          1. Launch the OnCall backend
        </p>
        <span
          class="root text text--secondary text--medium"
        >
          Run hobby, dev or production backend. See
           
          <a
            href="https://github.com/grafana/oncall#getting-started"
            rel="noreferrer"
            target="_blank"
          >
            <span
              class="root text text--link text--medium"
            >
              here
            </span>
          </a>
           
          on how to get started.
        </span>
      </div>
      <div
        class="info-block"
      >
        <p>
          2. Let us know the base URL of your OnCall API
        </p>
        <span
          class="root text text--secondary text--medium"
        >
          The OnCall backend must be reachable from your Grafana installation. Some examples are:
          <br />
          - http://host.docker.internal:8080
          <br />
          - http://localhost:8080
        </span>
      </div>
      <div
        class="css-1rplq84"
      >
        <div
          class="css-l4ykjo-Label"
        >
          <label>
            <div
              class="css-70qvj9"
            >
              OnCall backend URL
            </div>
          </label>
        </div>
        <div>
          <div>
            <div
              class="css-1cvb0sk-input-wrapper"
              data-testid="input-wrapper"
            >
              <div
                class="css-10lnb82-input-inputWrapper"
              >
                <input
                  class="css-8tk2dk-input-input"
                  data-testid="onCallApiUrl"
                  name="onCallApiUrl"
                />
              </div>
            </div>
          </div>
        </div>
      </div>
      <pre>
        <span
          class="root text text--link text--medium"
        >
          ohhh nooo there was an error from the OnCall API
        </span>
      </pre>
      <div
        class="root root--withBackGround info-block"
      >
        <span
          class="root text text--secondary text--medium"
        >
          Need help?
          <br />
          - Reach out to the OnCall team in the
           
          <a
            href="https://grafana.slack.com/archives/C02LSUUSE2G"
            rel="noreferrer"
            target="_blank"
          >
            <span
              class="root text text--link text--medium"
            >
              #grafana-oncall
            </span>
          </a>
           
          community Slack channel
          <br />
          - Ask questions on our GitHub Discussions page
           
          <a
            href="https://github.com/grafana/oncall/discussions/categories/q-a"
            rel="noreferrer"
            target="_blank"
          >
            <span
              class="root text text--link text--medium"
            >
              here
            </span>
          </a>
           
          <br />
          - Or file bugs on our GitHub Issues page
           
          <a
            href="https://github.com/grafana/oncall/issues"
            rel="noreferrer"
            target="_blank"
          >
            <span
              class="root text text--link text--medium"
            >
              here
            </span>
          </a>
        </span>
      </div>
      <button
        class="css-td06pi-button"
        type="submit"
      >
        <span
          class="css-1riaxdn"
        >
          Connect
        </span>
      </button>
    </form>
  </div>
</body>
`;
